### PR TITLE
issue 48 part 2: compiler and ir 

### PR DESCRIPTION
**Pull request by qwang98**
PR'ed against another branch (issue 48 part 1).

IR didn't need to change, because all signals are compiled into ir::Column and Placement, which is defined in compiler/cell manager.

### CODE REVIEW

File: src/compiler/cell_manager.rs
Comment (qwang98) [inline src/compiler/cell_manager.rs:213]: `signal_len_by_step_type` should be `step_height_by_step_type`. Sorry for the confusion.

File: src/compiler/cell_manager.rs
Comment (qwang98) [inline src/compiler/cell_manager.rs:340]: Let me know if you want me to do this check at the front (so that I delete these comments) or at the back (so that I uncomment these).

File: src/dsl.rs
Comment (qwang98) [inline src/dsl.rs:46]: Made these changes in this PR, so that I only need to merge in Part 3.

**Comment by qwang98** [comment]
Not sure why CI is not running, but I ran clippy and rustfmt. They can be fixed in the last part though.

File: src/dsl.rs
Comment (leolara) [reply src/dsl.rs:46]: I don't understand this sentence. Why is this showing this up if this is from part 1.

File: src/compiler/cell_manager.rs
Comment (leolara) [inline src/compiler/cell_manager.rs:200]: So for this CM specifically I think it is better to panic always if there are shared columns. We could revisit this in the future

**Review by leolara — CHANGES_REQUESTED** [review]
A quick review, 

After placement, the compiler should check that all steps are the same height if shared are not empty, we cannot leave that to the Cell Manager implementer, that could be someone else.

Change MaxWidthCellManager to always panic if there are shared signals.

File: src/dsl.rs
Comment (qwang98) [reply src/dsl.rs:46]: Apologies! I merged part 1 to main before fixing this so that I can merge main in part 2 rather than asking you to approve changes in part 1 again. This way it's quicker. However, I think you are right that for code organization reason it's the best to keep it in part 1.

File: src/compiler/cell_manager.rs
Comment (qwang98) [reply src/compiler/cell_manager.rs:200]: Fixed. Removed associated code for shared columns and panic at the top if there are shared columns.

**Comment by qwang98** [comment]
> A quick review,
> 
> After placement, the compiler should check that all steps are the same height if shared are not empty, we cannot leave that to the Cell Manager implementer, that could be someone else.
> 
> Change MaxWidthCellManager to always panic if there are shared signals.

Fixed!

**Comment by qwang98** [comment]
Ready for review again.

**Review by leolara — APPROVED** [review]

**Pull request merged**
